src/modelfree/common/utils.py

```diff
@@ -156,6 +171,104 @@ def make_session(graph=None):
     return sess
 
 
+class TrajectoryRecorder(VecMultiWrapper):
+    """Class for recording and saving trajectories in numpy.npz format.
+    For each episode, we record observations, actions, rewards and optionally network activations
+    for the agents specified by agent_indices.
+
+    :param venv: (VecEnv) environment to wrap
+    :param agent_indices: (list,int) indices of agents whose trajectories to record
+    """
+
+    def __init__(self, venv, agent_indices=None):
+        super().__init__(venv)
+
+        if agent_indices is None:
+            self.agent_indices = range(self.num_agents)
+        elif isinstance(agent_indices, int):
+            self.agent_indices = [agent_indices]
+
+        self.traj_dicts = [[defaultdict(list) for _ in range(self.num_envs)]
+                           for _ in self.agent_indices]
+        self.full_traj_dicts = [defaultdict(list) for _ in self.agent_indices]
+        self.prev_obs = None
+        self.actions = None
+
+    def step_async(self, actions):
+        self.actions = actions
+        self.venv.step_async(actions)
+
+    def step_wait(self):
+        observations, rewards, dones, infos = self.venv.step_wait()
+        self.record_traj(self.prev_obs, self.actions, rewards, dones, infos)
+        self.prev_obs = observations
+        return observations, rewards, dones, infos
+
+    def reset(self):
+        observations = self.venv.reset()
+        self.prev_obs = observations
+        return observations
+
+    def record_traj(self, prev_obs, actions, rewards, dones, infos):
+        """Record observations, actions, rewards, and (optionally) network activations
+        of one timestep in dict for current episode. Completed episode trajectories are
+        collected in a list in preparation for being saved to disk.
+
+        :param prev_obs: (np.ndarray<float>) observations from previous timestep
+        :param actions: (np.ndarray<float>) actions taken after observing prev_obs
+        :param rewards: (np.ndarray<float>) rewards from actions
+        :param dones: ([bool]) whether episode ended (not recorded)
+        :param infos: ([dict]) dicts with network activations if networks are transparent
+        :return: None
+        """
+        data_keys = ('observations', 'actions', 'rewards')
+        data_vals = (prev_obs, actions, rewards)
+        transparency_keys = ('ff', 'hid')  # we already record observations
```

Comment:
This is a good suggestion. I've attempted to implement this feature, was this what you had in mind?

Note I've renamed `transparency_keys` to `info_keys` since this seems more descriptive of what `TrajectoryRecorder` is actually doing. We just happen to be using it for transparency, but we could stuff other values in the `info` dict in the future.